test/test_type_promotion.py

```diff
@@ -430,6 +431,123 @@ def test_transpose(self, device):
         a = torch.tensor([[True, True], [False, True]], device=device)
         self.assertEqual(a.t() == 0, a.t() == False)  # noqa: E712
 
+    def _test_sparse_op_input_tensors(self, device, dtype, coalesced, zeros=True):
+        t = self._get_test_tensor(device, dtype, not zeros)
+        if zeros and dtype != torch.bool:
+            # ensure sparsity. Bool should already have sufficient sparsity.
+            mask = self._get_test_tensor(device, torch.bool)
+            t = t * mask
+        if coalesced:
+            s = t.to_sparse()
+        else:
+            s = t.to_sparse()
+            indices = torch.cat((s.indices(), s.indices()), 1)
+            values = torch.cat((s.values(), s.values()), 0)
+            s = torch.sparse_coo_tensor(indices=indices, values=values, size=s.size(), dtype=dtype, device=device)
+            t = s.to_dense()
+        self.assertEqual(s.is_coalesced(), coalesced)
+        self.assertEqual(s.dtype, dtype)
+        self.assertEqual(t.dtype, s.dtype)
+        return (t, s)
+
+    def _get_precision(self, dtype, coalesced):
+        if dtype == torch.half and not coalesced:
+            # very low precision for uncoalesced float16 sparse tensors since
+            # ops like (s1 + s2).to_dense() will add four low-precision
+            # floating point values.
+            return 5e-2
+        if dtype == torch.half:
+            return 1e-3
+        return 1e-5
+
+    def _test_sparse_op(self, op_name, inplace, dtype1, dtype2, device, coalesced):
+        suffix = '_' if inplace else ''
+        err = "{} {}({}, {})".format("  coalesced" if coalesced else "uncoalesced", op_name + suffix, dtype1, dtype2)
+
+        def op(t1, t2):
+            return getattr(t1, op_name + suffix)(t2)
+
+        add_sub = op_name == 'add' or op_name == 'sub'
+
+        (dense1, sparse1) = self._test_sparse_op_input_tensors(device, dtype1, coalesced)
+        (dense2, sparse2) = self._test_sparse_op_input_tensors(device, dtype2, coalesced, op_name != 'div')
+        common_dtype = torch.result_type(dense1, dense2)
+        if self.device_type == 'cpu' and common_dtype == torch.half:
+            self.assertRaises(RuntimeError, lambda: op(s1, d2))
+
+        if inplace and not torch.can_cast(common_dtype, dtype1):
+            self.assertRaises(RuntimeError, lambda: op(dense1, sparse2))
+            self.assertRaises(RuntimeError, lambda: op(sparse1, sparse2))
+            self.assertRaises(RuntimeError, lambda: op(sparse1, dense2))
```

Comment:
for (2), I think we'll need to refactor testing code around this more generally. There's precision logic elsewhere that doesn't work very well, often related to 'half' in particular, and a number of workarounds. e.g. `@precisionOverride` is sometimes used for handling half types, but is tied to the `@dtypes` annotation that wouldn't work for type promoting tests. If most cases are for half anyway, the default precision for half tests should just be lower at the source. Fixing it and consolidating the code everywhere seems outside the scope of this PR, and could be a large change, but I think that'll need to happen eventually.

(3) I'll try the smaller test tensors, which may help a bit. I think we can continue to iterate too though as we think of better approaches.

Reply:
(1) is tricky since this is short-circuiting cases that are invalid regardless of the operator or combination of dense/sparse. For operations that aren't supported at all, an error will be thrown before reaching the casting error.

Reply:
added a comment to try to clarify, not sure if you're happy enough with that for now. If I do another iteration with an explicit list of failure conditions, that might make this clearer too and easier to add the specific message without scrambling the rest of the test

Reply:
You're right, @nairbv, we do need to get better precision handling (and that is outside the scope of this PR). Type promotion tests are particularly tricky since they inherently handle multiple dtypes.